prevent reading directories as files when building fs cache #16417

jeremybobbin opened the pull request:
fixes #16417
check to see if path passed into `FileSystemInfo._readFileHash` is actually a file

**What kind of change does this PR introduce?**
bugfix

**Did you add tests for your changes?**
no

**Does this PR introduce a breaking change?**
no

**What needs to be documented once your changes are merged?**
nothing

alexander-akait reviewed (changes requested):
Please add a real reproducible test

jeremybobbin commented:
Further investigation leads me to conclude that this is not a webpack bug. This is a NodeJS bug, introduced in v23.0.0 & fixed in v23.2.0.

Introduced here:
https://github.com/nodejs/node/pull/54160

Fixed here:
https://github.com/nodejs/node/pull/55527

When calling `fs.readFile` on a path with a trailing slash("/path/"), it would throw an error & not call the provided callback.

```
fs.readFile(path[, options], callback)
```

I probably should have tried LTS node before reporting this bug.

alexander-akait commented:
@jeremybobbin yes, known issue, reported by me :smile: Anyway thank you, let's close
